packages/core-bridge/src/helpers/abort_controller.rs

```diff
@@ -0,0 +1,153 @@
+    // sending the abort signal. The good news is that nothing bad will happen if we call the JS
+    // abort callback multiple times.
+    js_counterpart: OnceLock<Arc<AbortControllerJsCounterpart>>,
+    aborted: OnceLock<String>,
+}
+
+struct AbortControllerJsCounterpart {
+    signal: Root<JsObject>,
+    abort: JsCallback<(String,), ()>,
+}
+
+impl AbortControllerInner {
+    /// Create the JS AbortController if it hasn't been created yet.
+    /// Returns a reference to the signal object that can be passed to JS.
+    fn ensure_js_initialized<'cx, C: Context<'cx>>(
+        &self,
+        cx: &mut C,
+    ) -> BridgeResult<Handle<'cx, JsValue>> {
+        if let Some(js_counterpart) = self.js_counterpart.get() {
+            // Already initialized, return the signal
+            return Ok(js_counterpart.signal.to_inner(cx).upcast());
+        }
+
+        // Not initialized yet, create the JS AbortController
+        let global = cx.global_object();
+        let abort_controller_class = global.get::<JsFunction, _, _>(cx, "AbortController")?;
+
+        let abort_controller = abort_controller_class.construct(cx, [])?;
+        let signal = abort_controller.get::<JsObject, _, _>(cx, "signal")?;
+        let abort_fn = abort_controller.get::<JsFunction, _, _>(cx, "abort")?;
+
+        let abort_cb = JsCallback::new(cx, abort_fn, Some(abort_controller));
+
+        let js_counterpart = Arc::new(AbortControllerJsCounterpart {
+            signal: signal.root(cx),
+            abort: abort_cb,
+        });
+
+        let js_counterpart = match self.js_counterpart.set(js_counterpart.clone()) {
+            Ok(()) => {
+                // If the Rust controller has already been aborted, call the JS abort callback now
+                // VALIDATE: Do we need a memory barrier here to ensure that js_counterpart and aborted are coherent?
+                //           I assume that the get() call ensures visibility of the js_counterpart
+                if let Some(aborted) = self.aborted.get() {
+                    // Fire and forget
+                    let _ = js_counterpart.abort.call_on_js_thread((aborted.clone(),));
+                }
+                js_counterpart
+            }
+            Err(js_counterpart) => js_counterpart,
+        };
+
+        Ok(js_counterpart.signal.to_inner(cx).upcast())
+    }
+
+    /// Immediately abort the `AbortController`, causing the JS side `signal` to fire.
+    fn abort(&self, reason: impl Into<String>) {
+        let reason = reason.into();
+        if let Ok(()) = self.aborted.set(reason.clone()) {
+            // If we haven't created the JS AbortController yet, there's nothing to abort
+            // VALIDATE: Do we need a memory barrier here to ensure that js_counterpart and aborted are coherent?
+            if let Some(js_counterpart) = self.js_counterpart.get() {
+                // Fire and forget
+                let _ = js_counterpart.abort.call_on_js_thread((reason.clone(),));
```

Comment:
Clone isn't necessary (clippy should whine about this, if it isn't we've probably got it set up wrong).